### PR TITLE
Enhance the inject_header function, fix bug in net/http server instru…

### PR DESCRIPTION
**Pull request by RonFed**
* Fix bug in net/http server instrumentation in which the span context was always generated as a random string rather than taking the parent one if exists
* Enhance the inject header function to be more readable

### CODE REVIEW

**Review by edeNFed — APPROVED** [review]

**Comment by RonFed** [comment]
@oxeye-gal Can you please have a look on the changes made to the inject_headers function ?

**Review by MikeGoldsmith — APPROVED** [review]

**Comment by MrAlias** [comment]
Please rebase these changes into `internal/pkg` (changed in https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/282). The `pkg` directory is going to be removed in https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/302 given that package is no longer used.

**Comment by RonFed** [comment]
> Please rebase these changes into `internal/pkg` (changed in #282). The `pkg` directory is going to be removed in #302 given that package is no longer used.

Rebased

File: internal/include/go_types.h [hunk 1]
```diff
@@ -54,6 +54,17 @@ struct map_bucket {
     void *overflow;
 };
 
+// A header for a Go map. Defined in go/src/runtime/map.go
+struct go_map_header {
```
Comment (MrAlias) [inline internal/include/go_types.h:58]: This seems brittle. What happens when the Go `hmap` field order or structure changes? Is there a better way to pass this from the Go memory space?

Comment (RonFed) [reply internal/include/go_types.h:58]: It looks like the last time a field was added was 6 years ago - so I think it is quite stable. Theoretically, we could do something similar to the offset tracker with this but I don't think it is worth the effort considering that it didn't change much

Comment (MrAlias) [reply internal/include/go_types.h:58]: What can be done with the offset tracker? I'd be more in favor of exploring less brittle options then assuming an unexported type won't change.

File: internal/include/go_types.h [hunk 2]
```diff
@@ -54,6 +54,17 @@ struct map_bucket {
...
+    u64 map_keyvalue_count;
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply internal/include/go_types.h:59]: What's the downside off using `s64` here instead?

File: internal/include/go_types.h (hunk 1 above)
Comment (RonFed) [reply internal/include/go_types.h:58]: I guess we could analyze the binary by looking for the map header struct and extracting the offsets of the fields that we care about.  In that case, we won't be able to read the entire header in one read operation as I did in this PR but would need to read one field at a time based on its offset.
As documented in go/src/runtime/map.go the count field that we read must be first, so I think we could be fine with just getting the offset of the `buckets` field which is not that bad. 
Again, I'm not sure if it is worth it. @edeNFed what do you think ?

File: internal/include/go_types.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (RonFed) [reply internal/include/go_types.h]: Done

File: internal/include/go_types.h (hunk 1 above)
Comment (MrAlias) [reply internal/include/go_types.h:58]: > so I think we could be fine with just getting the offset of the `buckets` field which is not that bad.

+1

I think generating these offsets should be similar to the other `runtime` offsets we generate in that they shouldn't be too hard to figure out (once we fix the offset generator).

Comment (RonFed) [reply internal/include/go_types.h:58]: @MrAlias Ok. Can I update the offsets json in this PR? or do you prefer for the offset-tracker to be fixed first?

Comment (RonFed) [reply internal/include/go_types.h]: I added the buckets field to the offset tracker and used the injected value to read the pointer

**Review by MrAlias — APPROVED** [review]

**Pull request merged**